make python remote exception to rethrow when using remote reference to itself

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29930 make python remote exception to rethrow when using remote reference to itself**

Right now, python call remote exception rethrown is coupled with deserializtiaon.
For owner ref, the setValue() and getValue() do not use serialization and deserialization, so when users create a ref to itself, and call ownerRef.to_here(), python call remote exception will not be rethrown.

This diff is to move remote exception rethrown out of deserialization, and exception can be handled for owner ref getValue()

close #29924

Differential Revision: [D18541916](https://our.internmc.facebook.com/intern/diff/D18541916/)

zhaojuanmao committed:
make python remote exception to rethrow when using remote reference to itself

Right now, python call remote exception rethrown is coupled with deserializtiaon.
For owner ref, the setValue() and getValue() do not use serialization and deserialization, so when users create a ref to itself, and call ownerRef.to_here(), python call remote exception will not be rethrown.

This diff is to move remote exception rethrown out of deserialization, and exception can be handled for owner ref getValue()

close #29924

Differential Revision: [D18541916](https://our.internmc.facebook.com/intern/diff/D18541916/)

[ghstack-poisoned]

pritamdamania87 commented:
Looks like we have a duplicate PR: https://github.com/pytorch/pytorch/pull/29929?

pritamdamania87 commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -199,11 +199,20 @@ template class UserRRef<py::object>;
-template <typename T>
-const T& OwnerRRef<T>::getValue() const {
+template <>
+const IValue& OwnerRRef<IValue>::getValue() const {
```
Why do we have to separate IValue and python objects like this? Shouldn't the same code work in python/torchscript mode since this is C++ code?

zhaojuanmao replied:
because python object and IValue exception handing is different per previous request. 

For IValue exception handling, it catches C++ exception and returns a message with message::Exception, then exception will be rethrown based on message::Exception type. 

For pyobject exception handing, it did not throw a C++ exception, but catch it in python side and wrap it into a RemoteException to caller. Caller will check this returned result is RemoteException or not and then rethrown. This is based on previous request that we need to return python call stack trace to users. C++ exception can only return a message, no stack trace

mrshenli commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -199,11 +199,20 @@ template class UserRRef<py::object>;
   // TODO: use callback to make this non-blocking
   std::unique_lock<std::mutex> lock(mutex_);
   valueCV_.wait(lock, [this] { return value_.has_value(); });
+  PythonRpcHandler::getInstance().handleException(value_.value());
```
Is this inconsistent with what we do in `PYTHON_CALL`? In `PYTHON_CALL` the exception is wrapped into payload in `generatePythonUDFResult`, and then the exception message is sent to the caller and thrown there in `toPyObjInternal`. But here, the exception is thrown on the callee (owner) and relies on `RequestCallback::operator()` to catch it and convert that into an exception message? Is this intentional?

zhaojuanmao replied:
both PYTHON_CALL and PYTHON_FETCH_CALL re-throw python exception during deserialization. 

Here, callee and caller are in the same process, did not go through deserialization, so skipped python exception re-thrown

mrshenli reviewed (commented):
Thanks for fixing! Shall we add tests for both `to_here` and `local_value` for both local (creating an `RRef` locally) and remote RRef?

zhaojuanmao commented:
@mrshenli and @pritamdamania let me move python exception rethrown to PyRRef::localValue() where only caller can call it.

zhaojuanmao committed:
Update on "make python remote exception to rethrow when using remote reference to itself"


Right now, python call remote exception rethrown is coupled with deserializtiaon.
For owner ref, the setValue() and getValue() do not use serialization and deserialization, so when users create a ref to itself, and call ownerRef.to_here(), python call remote exception will not be rethrown.

This diff is to move remote exception rethrown out of deserialization, and exception can be handled for owner ref getValue()

close #29924

Differential Revision: [D18541916](https://our.internmc.facebook.com/intern/diff/D18541916/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -77,6 +77,7 @@ py::object PyRRef::localValue() {
     const py::object& value =
         std::dynamic_pointer_cast<OwnerRRef<py::object>>(rref_)->getValue();
 
+    PythonRpcHandler::getInstance().handleException(value);
```
UserRRef.toHere() raises the error. Is it because this line?

https://github.com/pytorch/pytorch/blob/63675b1969d5437384851e0e543b2fe9785334ce/torch/csrc/distributed/rpc/rref.cpp#L170

zhaojuanmao replied:
no, it is because I added exception in get_value() which could be called in server side:(. 

I updated diff to add exception in localValue() in python/caller side only

mrshenli reviewed (approved):
LGTM! Please make sure all tests pass before landing.

mrshenli commented:
@zhaojuanmao please update the PR/diff description accordingly.

zhaojuanmao committed:
Update on "make python remote exception to rethrow when using remote reference to itself"


Right now, python call remote exception rethrown is coupled with deserializtiaon.
For owner ref, the setValue() and getValue() do not use serialization and deserialization, so when users create a ref to itself, and call ownerRef.to_here(), python call remote exception will not be rethrown.

This diff is to move remote exception rethrown out of deserialization, and exception can be handled for owner ref getValue()

close #29924

Differential Revision: [D18541916](https://our.internmc.facebook.com/intern/diff/D18541916/)

[ghstack-poisoned]

zhaojuanmao commented:
checked these tests actually finished, but CI showed they are pending

zhaojuanmao commented:
@pytorchbot retest please

zhaojuanmao committed:
Update on "make python remote exception to rethrow when using remote reference to itself"


Right now, python call remote exception rethrown is coupled with deserializtiaon.
For owner ref, the setValue() and getValue() do not use serialization and deserialization, so when users create a ref to itself, and call ownerRef.to_here(), python call remote exception will not be rethrown.

This diff is to move remote exception rethrown out of deserialization, and exception can be handled for owner ref getValue()

close #29924

Differential Revision: [D18541916](https://our.internmc.facebook.com/intern/diff/D18541916/)

[ghstack-poisoned]

zzzwen commented on torch/distributed/rpc/internal.py:
```diff
@@ -119,6 +119,9 @@ def _run_function(binary_data, tensor_table):
     return result
 
+def _handle_exception(result):
+    if isinstance(result, RemoteException):
```
question/nit, do we lost the original Exception type while wrapping it as `RemoteException`? If we are able to retrieve the original type, rethrowing it as the original type will be helpful.

zhaojuanmao replied:
@zzzwen all exception type, msg and stack trace are wrapped into remoteException and users can read these info in client.

zhaojuanmao commented:
some tests stuck again, rerun workflow did not trigger to retest it somehow.

for safety, I'm rebasing it and try one more round of tests again.

zhaojuanmao committed:
Update on "make python remote exception to rethrow when using remote reference to itself"


Right now, python call remote exception rethrown is coupled with deserializtiaon.
For owner ref, the setValue() and getValue() do not use serialization and deserialization, so when users create a ref to itself, and call ownerRef.to_here(), python call remote exception will not be rethrown.

This diff is to move remote exception rethrown out of deserialization, and exception can be handled for owner ref getValue()

close #29924

Differential Revision: [D18541916](https://our.internmc.facebook.com/intern/diff/D18541916/)

[ghstack-poisoned]
